apps/emqx_rule_engine/src/emqx_rule_engine_jwt_worker.erl

```diff
@@ -0,0 +1,220 @@
+        , handle_info/2
+        , format_status/1
+        , format_status/2
+        ]).
+
+-include_lib("jose/include/jose_jwk.hrl").
+-include_lib("emqx_rule_engine/include/rule_engine.hrl").
+-include_lib("emqx_rule_engine/include/rule_actions.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-type config() :: #{ private_key := binary()
+                   , resource_id := resource_id()
+                   , expiration := timer:time()
+                   , table := ets:table()
+                   , iss := binary()
+                   , sub := binary()
+                   , aud := binary()
+                   , kid := binary()
+                   , alg := binary()
+                   }.
+-type jwt() :: binary().
+-type state() :: #{ refresh_timer := undefined | timer:tref()
+                  , resource_id := resource_id()
+                  , expiration := timer:time()
+                  , table := ets:table()
+                  , jwt := undefined | jwt()
+                    %% only undefined during startup
+                  , jwk := undefined | jose_jwk:key()
+                  , iss := binary()
+                  , sub := binary()
+                  , aud := binary()
+                  , kid := binary()
+                  , alg := binary()
+                  }.
+
+-define(refresh_jwt, refresh_jwt).
+
+%%-----------------------------------------------------------------------------------------
+%% API
+%%-----------------------------------------------------------------------------------------
+
+-spec start_link(config(), reference()) -> gen_server:start_ret().
+start_link(#{ private_key := _
+            , expiration := _
+            , resource_id := _
+            , table := _
+            , iss := _
+            , sub := _
+            , aud := _
+            , kid := _
+            , alg := _
+            } = Config,
+           Ref) ->
+    gen_server:start_link(?MODULE, {Config, Ref}, []).
+
+-spec lookup_jwt(resource_id()) -> {ok, jwt()} | {error, not_found}.
+lookup_jwt(ResourceId) ->
+    ?MODULE:lookup_jwt(?JWT_TABLE, ResourceId).
+
+-spec lookup_jwt(ets:table(), resource_id()) -> {ok, jwt()} | {error, not_found}.
+lookup_jwt(TId, ResourceId) ->
+    try
+        case ets:lookup(TId, {ResourceId, jwt}) of
+            [{{ResourceId, jwt}, JWT}] ->
+                {ok, JWT};
+            [] ->
+                {error, not_found}
+        end
+    catch
+        error:badarg ->
+            {error, not_found}
+    end.
+
+%%-----------------------------------------------------------------------------------------
+%% gen_server API
+%%-----------------------------------------------------------------------------------------
+
+-spec init({config(), Ref}) -> {ok, state(), {continue, {make_key, binary(), Ref}}}
+                               | {stop, {error, term()}}
+              when Ref :: reference().
+init({#{private_key := PrivateKeyPEM} = Config, Ref}) ->
+    {ok, _} = application:ensure_all_started(jose),
```

Comment:
If jose is not already started, then I wonder why this does NOT cause application controller dead-lock.
that is, starting the supervision tree is a sync-call from app controller, and making another call to itself is a dead-lock?

Reply:
I saw some similar things in other rule actions, so I did the same here. :see_no_evil: 

The app controller doesn't spawn any workers, only the JWT supervisor. The workers are spawned on-demand by the rule actions.

Reply:
I'll add as an app dependency rather than start it here.

Reply:
:heavy_check_mark: 

https://github.com/emqx/emqx/pull/9241/commits/f158315812ebbcc3f1ef5c17cbd8f65db477557c